У меня не работал модуль multiprocessing, но оказалось, что в юпитере с этим модулем вообще проблемы. Поэтому я установил адаптированную для юпитера версию  multiprocess

In [1]:
from Bio import SeqIO
import multiprocess 
import multiprocessing
import time

In [2]:
def parse_sequence(record_id, record_seq):
    parser = {}
    for word in record_seq:
        if word in parser:
            parser[word] += 1
        else:
            parser[word] = 1
    print(f"Contig {record_id}:    ", end = "")
    parsered = []
    for key, value in zip(parser.keys(), parser.values()):
            parsered.append(f"{key}={value}")
    print(*parsered, sep = ", ")
    return 

Сначала вариант реализации в лоб (то, что первое в голову приходит)

In [3]:
def parse_fasta(filename, threads):
    records = SeqIO.parse(filename, "fasta")
    selected_records = []
    for record in records:
        selected_records.append(record)
        if len(selected_records) == threads:
            processes = []
            for record in selected_records:
                processes.append(multiprocess.Process(target=parse_sequence, args=(record.id, record.seq)))
            for proc in processes:
                proc.start()
            for proc in processes:
                proc.join()
            selected_records = []
    if len(selected_records) > 0: # take last ones
        processes = []
        for record in selected_records:
            processes.append(multiprocess.Process(target=parse_sequence, args=(record.id, record.seq)))
        for proc in processes:
            proc.start()
        for proc in processes:
            proc.join()    

In [22]:
start = time.time()
parse_fasta("contigs.fna", 1)
end = time.time()
print(f"executing time is {end-start}")

Contig NC_000001.11:    N=18475408, t=25922550, a=25810086, c=16788841, g=16855483, C=31266202, G=31256045, T=41321614, A=41260191, M=1, R=1
Contig NT_187361.1:    G=21162, A=27812, T=28097, C=22678, g=14746, t=22941, a=22589, c=15030
Contig NT_187362.1:    A=3698, G=2858, T=3012, C=2663, t=4874, c=4118, a=6741, g=4068
Contig NT_187363.1:    g=15374, a=21515, t=22087, c=14581, G=11355, C=11061, T=16322, A=15387
Contig NT_187364.1:    g=6917, a=13516, t=14842, c=8228, G=4664, A=6616, T=7146, C=4931
Contig NT_187365.1:    A=6892, C=4131, T=7573, G=4071, t=6280, g=3102, c=3019, a=5108
Contig NT_187366.1:    t=4163, g=2926, a=4770, c=3071, C=6163, T=7662, A=7516, G=5939
Contig NT_187367.1:    G=23619, A=25329, T=25118, C=22712, a=22014, g=16992, t=23518, c=16741
Contig NT_187368.1:    G=7661, A=7857, C=7266, T=7459, g=2249, c=2243, a=3157, t=2853
Contig NT_187369.1:    C=8914, T=6771, A=6735, a=2757, g=2497, t=2655, c=2626, G=8762
Contig NC_000002.12:    N=1645292, C=32288347, G=32342686, 

Contig NT_187513.1:    G=20091, A=29586, T=30031, C=19458, a=27353, c=16483, t=27453, g=16284
Contig NT_167211.2:    c=33659, a=26477, t=42902, g=13759, A=15750, T=17305, C=14400, G=12120, N=236
Contig NT_113889.1:    G=20430, A=25697, T=25901, C=20062, a=20333, c=13234, t=22146, g=13344
Contig NW_012132914.1:    T=73335, C=63977, G=63496, A=72771, g=42795, c=42977, a=53524, t=54268
Contig NW_015495298.1:    G=38466, A=43933, T=44635, C=38387, g=25207, t=32084, a=30952, c=24995
Contig NW_025791756.1:    g=124667, a=150112, t=151157, c=124447, G=256685, T=254754, C=256170, A=254694
Contig NW_011332688.1:    C=51383, A=48885, G=52585, T=47437, c=24357, t=26916, a=28635, g=25344
Contig NW_014040926.1:    G=46272, A=55835, T=53274, c=33911, g=33383, t=37584, a=40236, C=49443
Contig NW_009646195.1:    a=21757, c=20849, g=20301, T=21129, C=18342, A=20547, t=23448, G=19827
Contig NW_018654706.1:    C=15846, T=20031, A=19506, G=15709, t=25256, g=22104, a=25212, c=22472
Contig NW_019805487.1:  

Contig NW_018654721.1:    G=13211, C=12850, T=19755, A=21495, t=28797, a=30413, c=22478, g=22799
Contig NW_011332701.1:    T=873796, G=716990, A=884485, C=720328, g=353224, t=525633, a=533071, c=354444, N=36991
Contig NW_021160018.1:    G=50220, C=49489, T=64177, A=63661, g=31870, a=51675, t=52420, c=33003
Contig NW_021160017.1:    c=437704, a=652048, t=677619, g=435775, A=926901, C=711545, G=716291, T=942566
Contig NW_012132920.1:    A=187178, C=144338, T=184379, G=143763, a=110816, g=71661, t=111702, c=71501, N=1340026
Contig NW_025791798.1:    C=28603, A=36261, G=27602, t=22494, g=15590, a=22348, c=16746, T=34363
Contig NW_021160016.1:    A=62817, C=71310, G=67230, T=60219, g=25075, a=27996, t=28044, c=26573
Contig NW_025791797.1:    a=118058, g=79091, t=129850, c=79949, A=202713, T=214645, C=163805, G=166463
Contig NW_021160015.1:    t=10518, g=5674, c=5571, a=9105, G=16201, A=17519, T=17939, C=15236
Contig NW_025791799.1:    T=11582, C=19074, A=11691, G=18551, g=6919, t=6743, c=72

Contig NT_187571.1:    C=69070, A=46445, G=67786, T=45437, g=18617, c=19013, t=20645, a=18828
Contig NT_187569.1:    C=20701, T=17692, G=18535, A=19184, g=12976, t=14505, a=15726, c=12925
Contig NW_003315928.1:    T=35875, C=23566, G=24965, A=33784, c=7917, a=13821, t=14888, g=8172
Contig NW_003315929.1:    A=13297, C=11825, T=12687, G=11905, t=7708, g=4581, a=5184, c=4364
Contig NW_003315930.1:    A=32601, C=24068, T=33505, G=24592, t=18908, c=10617, g=10843, a=16152
Contig NW_003315931.1:    T=9711, A=10373, C=7669, G=7639, g=5416, t=6624, a=7686, c=4914
Contig NT_187578.1:    c=23989, t=44716, a=46225, g=25051, G=60486, T=89794, C=58733, A=90088
Contig NW_003315934.1:    C=23468, T=34762, A=34310, G=23662, t=19495, g=11808, a=19530, c=12219
Contig NT_187579.1:    a=29234, t=34214, g=18447, c=19601, T=22550, C=16994, A=23100, G=17356
Contig NW_003315935.1:    G=38120, A=48976, T=45600, C=38909, a=45640, c=27800, t=37426, g=27331
Contig NT_187586.1:    C=42074, T=28140, G=44578, A=279

Contig NT_187676.1:    a=20646, g=17974, t=20946, c=15309, G=28321, T=28621, A=29784, C=26334
Contig NT_187677.1:    C=25755, A=29567, G=27558, T=28309, t=21821, g=18201, a=22421, c=15720
Contig NT_187683.1:    A=29538, C=26085, T=28354, G=27994, a=20496, g=17701, c=15232, t=20803
Contig NT_187684.1:    T=35971, C=34327, A=34104, G=31940, c=14392, t=18871, g=12867, a=18301
Contig NT_187685.1:    t=16722, c=12309, g=11263, a=16074, C=28284, A=28483, T=30250, G=26763
Contig NT_187686.1:    T=38639, G=34303, A=36502, C=36610, t=20319, c=15527, a=19781, g=14051
Contig NT_187687.1:    T=30337, C=28463, A=28573, G=26822, c=12403, t=16710, g=11190, a=16039
Contig NT_113949.2:    G=28410, A=30286, T=31982, C=30314, t=16854, c=12429, a=15759, g=11347
Contig NC_012920.1:    G=2169, A=5124, T=4094, C=5181, N=1
executing time is 1617.0885450839996


In [23]:
start = time.time()
parse_fasta("contigs.fna", 2)
end = time.time()
print(f"executing time is {end-start}")

Contig NT_187362.1:    A=3698, G=2858, T=3012, C=2663, t=4874, c=4118, a=6741, g=4068
Contig NT_187363.1:    g=15374, a=21515, t=22087, c=14581, G=11355, C=11061, T=16322, A=15387
Contig NT_187364.1:    g=6917, a=13516, t=14842, c=8228, G=4664, A=6616, T=7146, C=4931
Contig NT_187365.1:    A=6892, C=4131, T=7573, G=4071, t=6280, g=3102, c=3019, a=5108
Contig NT_187366.1:    t=4163, g=2926, a=4770, c=3071, C=6163, T=7662, A=7516, G=5939
Contig NT_187367.1:    G=23619, A=25329, T=25118, C=22712, a=22014, g=16992, t=23518, c=16741
Contig NT_187368.1:    G=7661, A=7857, C=7266, T=7459, g=2249, c=2243, a=3157, t=2853
Contig NT_187369.1:    C=8914, T=6771, A=6735, a=2757, g=2497, t=2655, c=2626, G=8762
Contig NC_000001.11:    N=6199831, t=8532143, a=8494139, c=5516190, g=5538716, C=10246586, G=10241008, T=13572650, A=13563057
Contig NT_187361.1:    G=1172, A=1375, T=1707, C=990, g=33798, t=50508, a=46205, c=26621
Contig NT_187371.1:    G=824, A=937, T=1236, C=690, a=43917, c=25879, t=48901, 

Contig NT_113889.1:    G=20430, A=25697, T=25901, C=20062, a=20333, c=13234, t=22146, g=13344
Contig NW_012132914.1:    T=73335, C=63977, G=63496, A=72771, g=42795, c=42977, a=53524, t=54268
Contig NW_015495298.1:    G=38466, A=43933, T=44635, C=38387, g=25207, t=32084, a=30952, c=24995
Contig NW_025791756.1:    g=124667, a=150112, t=151157, c=124447, G=256685, T=254754, C=256170, A=254694
Contig NW_011332688.1:    C=51383, A=48885, G=52585, T=47437, c=24357, t=26916, a=28635, g=25344
Contig NW_014040926.1:    G=46272, A=55835, T=53274, c=33911, g=33383, t=37584, a=40236, C=49443
Contig NW_009646195.1:    a=21757, c=20849, g=20301, T=21129, C=18342, A=20547, t=23448, G=19827
Contig NW_018654706.1:    C=15846, T=20031, A=19506, G=15709, t=25256, g=22104, a=25212, c=22472
Contig NW_019805487.1:    A=8659, T=9530, C=8803, G=9559, g=1677, t=2253, a=2426, c=2048
Contig NW_009646194.1:    C=35453, A=32348, G=36974, T=35487, a=11448, t=14514, g=10687, c=9583
Contig NW_018654707.1:    G=13887,

Contig NW_021160018.1:    G=50220, C=49489, T=64177, A=63661, g=31870, a=51675, t=52420, c=33003
Contig NW_021160017.1:    c=437704, a=652048, t=677619, g=435775, A=926901, C=711545, G=716291, T=942566
Contig NW_012132920.1:    A=187178, C=144338, T=184379, G=143763, a=110816, g=71661, t=111702, c=71501, N=1340026
Contig NW_025791798.1:    C=28603, A=36261, G=27602, t=22494, g=15590, a=22348, c=16746, T=34363
Contig NW_021160016.1:    A=62817, C=71310, G=67230, T=60219, g=25075, a=27996, t=28044, c=26573
Contig NW_025791797.1:    a=118058, g=79091, t=129850, c=79949, A=202713, T=214645, C=163805, G=166463
Contig NW_021160015.1:    t=10518, g=5674, c=5571, a=9105, G=16201, A=17519, T=17939, C=15236
Contig NW_025791799.1:    T=11582, C=19074, A=11691, G=18551, g=6919, t=6743, c=7209, a=5946
Contig NW_013171812.1:    G=11656, T=9042, C=11552, A=9843, c=3853, a=4970, g=4222, t=3878
Contig NW_019805500.1:    T=87804, G=74248, A=90019, C=77052, t=45241, g=31506, a=41803, c=32742
Contig NW_01

Contig NT_187571.1:    C=20594, A=19082, G=18636, T=17386, g=12716, c=12680, t=14246, a=15411
Contig NT_187574.1:    A=45971, T=45094, G=67224, C=68202, t=20648, g=18642, a=18845, c=19002
Contig NT_187569.1:    C=12979, T=13372, G=12655, A=13896Contig NW_003315928.1:    , T=35875g=4830, C=23566, t=7935, , G=24965a=5465, , A=33784c=4622, 
c=7917, a=13821, t=14888, g=8172
Contig NW_003315929.1:    A=10596, C=7837, T=9823, G=7797, t=6523, g=5354, a=7644, c=4812
Contig NW_003315930.1:    A=32601, C=24068, T=33505, G=24592, t=18908, c=10617, g=10843, a=16152
Contig NW_003315931.1:    T=34290, A=33808, C=23255, G=23462, g=11622, t=19246, a=19236, c=12056
Contig NT_187578.1:    c=23989, t=44716, a=46225, g=25051, G=60486, Contig NT_187579.1:    T=89794a=29234, , C=58733t=34214, , A=90088
g=18447, c=19601, T=22550, C=16994, A=23100, G=17356
Contig NW_003315935.1:    G=44558, A=28592, T=28613, C=42007, a=17185, c=17411, t=16536, g=17700
Contig NW_003315934.1:    C=38486, T=45147, A=48436, G=374

Contig NT_187675.1:    C=26046, A=29523, G=27839, T=28313, t=21047, c=15455, g=18013, a=20877
Contig NT_187676.1:    a=22528, g=18372, t=21889, c=15768, G=28043, T=28710, A=29956, C=26068
Contig NT_187677.1:    C=25322, A=28630, G=27098, T=27664, t=20451, g=17347, a=20019, c=14978
Contig NT_187683.1:    A=34356, C=34596, T=36002, G=32269, a=18012, g=12853, c=14254, t=18681
Contig NT_187684.1:    T=30254, C=28288, A=28484, G=26763, c=12018, t=16438, g=10950, a=15589
Contig NT_187685.1:    t=20416, c=15596, g=14149, a=19905, C=36604, Contig NT_187686.1:    A=36499T=31007, T=38639, , G=27483G=34300
, A=29244, C=29044, t=17384, c=12991, a=17113, g=11775
Contig NT_113949.2:    G=3143, A=6158, T=5307, C=6186, t=941, c=723, a=929, g=711, N=1
Contig NT_187687.1:    T=30785, C=29289, A=29239, G=27455, c=11739, t=15982, g=10659, a=14905
Contig NC_012920.1:    G=66, A=121, T=111, C=117
executing time is 855.4581570625305


Теперь более сложная реализация

In [4]:
class MyProcess(multiprocess.Process):
    def __init__(self, records):
        super().__init__()
        self.records = records
    
    def run(self):
        for record in self.records:
            parse_sequence(record.id, record.seq)

In [5]:
def parse_fasta_fast(filename, threads):
    records = SeqIO.parse(filename, "fasta")
    processes = []
    for i in range(threads):
        processes.append(MyProcess(records))
    for proc in processes:
        proc.start()
    for proc in processes:
        proc.join()

In [26]:
start = time.time()
parse_fasta_fast("contigs.fna", 2)
end = time.time()
print(f"executing time is {end-start}")

Contig NT_187361.1:    G=8575, A=11613, T=11678, C=9321, g=5560, t=8349, a=9057, c=5719
Contig NT_187362.1:    A=15232, G=11637, T=16356, C=11140, t=21949, c=14588, a=21212, g=15356
Contig NT_187364.1:    g=6917, a=13516, t=14842, c=8228, G=4664, A=6616, T=7146, C=4931
Contig NT_187365.1:    A=6892, C=4131, T=7573, G=4071, t=6280, g=3102, c=3019, a=5108
Contig NT_187366.1:    t=4163, g=2926, a=4770, c=3071, C=6163, T=7662, A=7516, G=5939
Contig NT_187367.1:    G=23619, A=25329, T=25118, C=22712, a=22014, g=16992, t=23518, c=16741
Contig NT_187368.1:    G=7661, A=7857, C=7266, T=7459, g=2249, c=2243, a=3157, t=2853
Contig NT_187369.1:    C=8914, T=6771, A=6735, a=2757, g=2497, t=2655, c=2626, G=8762
Contig NC_000001.11:    N=6184884, t=8443904, a=8427671, c=5463204, g=5494449, C=10161771, G=10172093, T=13450969, A=13455083, R=1
Contig NT_187363.1:    g=33344, a=45931, t=49817, c=26230, G=411, C=343, T=601, A=644
Contig NT_187371.1:    G=824, A=937, T=1236, C=690, a=43917, c=25879, t=489

Contig NT_113889.1:    G=20430, A=25697, T=25901, C=20062, a=20333, c=13234, t=22146, g=13344
Contig NW_012132914.1:    T=73335, C=63977, G=63496, A=72771, g=42795, c=42977, a=53524, t=54268
Contig NW_015495298.1:    G=38466, A=43933, T=44635, C=38387, g=25207, t=32084, a=30952, c=24995
Contig NW_025791756.1:    g=124667, a=150112, t=151157, c=124447, G=256685, T=254754, C=256170, A=254694
Contig NW_011332688.1:    C=51383, A=48885, G=52585, T=47437, c=24357, t=26916, a=28635, g=25344
Contig NW_014040926.1:    G=46272, A=55835, T=53274, c=33911, g=33383, t=37584, a=40236, C=49443
Contig NW_009646195.1:    a=21757, c=20849, g=20301, T=21129, C=18342, A=20547, t=23448, G=19827
Contig NW_018654706.1:    C=15846, T=20031, A=19506, G=15709, t=25256, g=22104, a=25212, c=22472
Contig NW_019805487.1:    A=8659, T=9530, C=8803, G=9559, g=1677, t=2253, a=2426, c=2048
Contig NW_009646194.1:    C=35453, A=32348, G=36974, T=35487, a=11448, t=14514, g=10687, c=9583
Contig NW_018654707.1:    G=13887,

Contig NW_021160014.1:    G=13811, A=22553, C=13379, T=21096, t=29572, g=22998, a=31109, c=22705
Contig NW_011332701.1:    T=65459, G=51739, A=65249, C=50950, g=32162, t=52715, a=52100, c=33141
Contig NW_018654721.1:    G=715642, C=719064, T=872753, A=883124, t=525359, a=532680, c=354313, g=352937, N=36991
Contig NW_021160017.1:    c=437704, a=652048, t=677619, g=435775, A=926901, C=711545, G=716291, T=942566
Contig NW_012132920.1:    A=35276, C=28039, T=33554, G=27022, a=21572, g=15155, t=21978, c=16258
Contig NW_021160016.1:    A=62817, C=71310, G=67230, T=60219, g=25075, a=27996, t=28044, c=26573
Contig NW_021160018.1:    G=143700, C=144269, T=184231, A=187100, g=71521, a=110630, t=111428, c=71389, N=1340026
Contig NW_025791798.1:    C=15923, A=18631, G=16880, t=10736, g=5906, a=9795, c=5759, T=18928
Contig NW_025791797.1:    a=118058, g=79091, t=129850, c=79949, A=202713, T=214645, C=163805, G=166463
Contig NW_025791799.1:    T=11582, C=19074, A=11691, Contig NW_021160015.1:    t=4

Contig NT_187571.1:    C=20594, A=19082, G=18636, T=17386, g=12716, c=12680, t=14246, a=15411
Contig NT_187574.1:    A=45971, T=45094, G=67224, C=68202, t=20648, g=18642, a=18845, c=19002
Contig NT_187569.1:    C=12979, Contig NW_003315928.1:    T=35875, T=13372C=23566, , G=24965G=12655, , A=33784A=13896, , c=7917, g=4830a=13821, , t=7935t=14888, , g=8172a=5465, 
c=4622
Contig NW_003315930.1:    A=10340, C=7602, T=9650, G=7435, t=7114, c=5379, g=5774, a=8237
Contig NW_003315929.1:    A=32857, C=24303, T=33678, G=24954, t=18317, g=10423, a=15559, c=10050
Contig NW_003315931.1:    T=34290, A=33808, C=23255, G=23462, g=11622, t=19246, a=19236, c=12056
Contig NT_187579.1:    a=29234, t=34214, Contig NT_187578.1:    g=18447c=23989, , t=44716c=19601, , a=46225T=22550, , g=25051C=16994, , G=60486A=23100, T=89794, , G=17356C=58733
, A=90088
Contig NW_003315934.1:    C=41584, T=28160, A=28052, G=43885, t=16244, g=17530, a=16740, c=17311
Contig NW_003315935.1:    G=38120, A=48976, T=45600, C=389

Contig NT_187675.1:    C=26046, A=29523, G=27839, T=28313, t=21047, c=15455, g=18013, a=20877
Contig NT_187676.1:    a=22528, g=18372, t=21889, c=15768, G=28043, T=28710, A=29956, C=26068
Contig NT_187677.1:    C=25322, A=28630, G=27098, T=27664, t=20451, g=17347, a=20019, c=14978
Contig NT_187683.1:    A=34356, C=34596, T=36002, G=32269, a=18012, g=12853, c=14254, t=18681
Contig NT_187684.1:    T=30254, C=28288, A=28484, G=26763, c=12018, t=16438, g=10950, a=15589
Contig NT_187685.1:    t=20416, c=15596, g=14149, a=19905, C=36604, A=36499, T=38639, G=34300
Contig NT_187686.1:    T=31007, G=27483, A=29244, C=29044, t=17384, c=12991, a=17113, g=11775
Contig NT_113949.2:    G=3143, A=6158, T=5307, C=6186, t=941, c=723, a=929, g=711, N=1
Contig NT_187687.1:    T=30785, C=29289, A=29239, G=27455, c=11739, t=15982, g=10659, a=14905
Contig NC_012920.1:    G=66, A=121, T=111, C=117
executing time is 802.1312289237976


Интересно то, что данная реализация отрабатывает нормально, то есть здесь отрабатывает один и тот же итератор в 2 процессах, повторения разных записей нет. Это не совсем вяжется с теорией на лекции, но это работает правильно

## Итоги 

* без распараллеливания: 1617 c
* базовая версия распараллеливания: 855 c
* продуманная версия распараллеливания: 802 с

В конечном итоге получили ускорение в 2 раза (как и ожидалось)